### PR TITLE
ui: Ember upgrade 2.18 > 3.12

### PR DESCRIPTION
**Pull request by johncowen**
This PR supercedes https://github.com/hashicorp/consul/pull/6317 and https://github.com/hashicorp/consul/pull/6406.

As `ember-cli-update --run-codemods` touches a lot of code (more or less the entire codebase), rebasing/merging the main branch to the above PRs to bring them into sync is far from straightforwards. We decided it would be much easier to re-run the automated ember update/codemods and then cherry-pick the manual amends required for the upgrade on top of that rather than sift through the conflicts.

This PR does that all in one shot (so 2.18 > 3.12).

Please see https://github.com/hashicorp/consul/pull/6317 and https://github.com/hashicorp/consul/pull/6406 for details on the more interesting amends required for the upgrade, although the cherry-picked commits here also include reasonably detailed information.

Finally, there is one additional change here, which we believe could be a consequence of a change to RSVP in the upgrade:

Part of the additional features that had been added since the above PRs revolved around adding support for visually highlighting the leader in the node listing. This requires an additional request to a different API endpoint on entering the node listing view (kicked off by `findByLeader`):

https://github.com/hashicorp/consul/blob/74461406e01ee7c73387bcf4b91bedad1802dadf/ui-v2/app/routes/dc/nodes/index.js#L16-L18

We have an acceptance tests which tests that when a user clicks a link it takes them to the correct page. It also currently asserts the _last_ GET request made to the API when the user arrives on that page.

https://github.com/hashicorp/consul/blob/ddca6b934a9b324d08b690bba44d34c67be83dcd/ui-v2/tests/acceptance/page-navigation.feature#L15-L31

 Previous to the upgrade the _last_ request was to the `status/leader` endpoint, after the upgrade the _last_ request changes to the `internal/ui/nodes` endpoint. We don't know exactly the reasons why the upgrade causes this change, but we are happy enough that it doesn't actually affect anything inportant. Therefore, this PR includes a change to the test to ensure the test passes (ddca6b9).

Notes:

1. We should probably change this so we are testing all the API endpoints called, without asserting an order (in all cases I can think of the order doesn't matter here, I would imagine that at least in most cases I can't think of - it probably doesn't matter). Basically, loosen up the test slightly so it's less brittle.
2. As the order doesn't matter here, I'm happy making this change and then addressing the test not to assert the order of things at a later date. This `.feature` file was one of the first we did to give us maximum coverage/confidence as quickly/easily as possible, so it probably needs completely breaking out now and merging into the more finer grained `.feature` files.

### CODE REVIEW

File: ui-v2/package.json [hunk 1]
```diff
@@ -3,24 +3,25 @@
...
   "directories": {
     "doc": "doc",
     "test": "tests"
   },
-  "repository": "",
   "scripts": {
     "build": "ember build --environment production",
     "build:staging": "ember build --environment staging",
-    "build-ci": "ember build --environment test",
+    "build:ci": "ember build --environment test",
```
Comment (gregone) [inline ui-v2/package.json]: the `GNUMakefile` still uses `yarn run build-ci`, do you need to update that too?

Comment (johncowen) [reply ui-v2/package.json]: Oh wow, such a good spot @gregone thanks! I've amended in a commit on the end (I just changed it the one that I left to `:`. Now I can see why it was duplicated in the cherrypick!

**Comment by johncowen** [comment]
Thanks again for looking at this @gregone 

Yeah so this is _mainly_ automated, but there are some manual changes that were already reviewed in the mentioned PRs, the only thing I added here was the test thing I mention in the PR description and then of course the 'makefile target de-duplication' 😆 (oops thanks again for that!) which I spotted post PR.

> One warning: watch out when removing ember-shims I think it caused some testing issues for us and I had to re-add it

This was removed by the automated upgrade itself, and tbh I didn't exactly know what it was without looking it up. Seems to be dealing with the old way of importing things in ember? In which case if all my tests pass all should be ok? Just incase though, can you remember how your problems manifested? I'm guessing "tests pass so, 'this is fine'"?

**Comment by gregone** [comment]
ha, found it. it was in the deprecations warnings, and it seems that there was an issue with ember-cli-flash, which you also seem to use https://github.com/hashicorp/terraform-registry/pull/459

**Comment by johncowen** [comment]
Cool thanks Greg, I took at look at our flashes and they seem to be fine 🤞 . It looks like `ember-cli-flash` only depends on `ember-shims` in its `devDependencies` so we don't pull it in from there. I also did a `yarn why ember-shims` and nothing came up.

Thanks again for the once over, I think I'm happy to merge this now - I also have https://github.com/hashicorp/consul/pull/6387 to redo over the top of this. FYI this plan is to keep this hanging in `ui-staging` for a good few weeks before we put it onto `master` so it will get plenty of non-automated tyre kicking before then.

**Pull request merged**